kong/runloop/handler.lua

```diff
@@ -832,6 +832,99 @@ local function set_init_versions_in_cache()
+    if upstream_http_connection ~= "keep-alive" and
+       upstream_http_connection ~= "close"      and
+       upstream_http_connection ~= "upgrade"
+    then
+      for _, header_name in csv(upstream_http_connection) do
+        if header_name ~= "close" and header_name ~= "upgrade" and header_name ~= "keep-alive" then
+          header[header_name] = nil
+        end
+      end
+    end
+
+    local upgrade = var.upstream_http_upgrade
+    if upgrade and lower(upgrade) ~= lower(var.upstream_upgrade) then
+      header["Upgrade"] = nil
+    end
+
+    -- remove trailer response header when client didn't ask for them
+    if var.upstream_te == "" and var.upstream_http_trailer then
+      header["Trailer"] = nil
+    end
+
+    local upstream_status_header = constants.HEADERS.UPSTREAM_STATUS
+    if kong.configuration.enabled_headers[upstream_status_header] then
+      local upstream_status = ctx.buffered_status or tostring(tonumber(sub(var.upstream_status or "", -3)))
+      header[upstream_status_header] = upstream_status
+      if not header[upstream_status_header] then
+        log(ERR, "failed to set ", upstream_status_header, " header")
+      end
+    end
+
+    -- if this is the last try and it failed, save its state to correctly log it
+    local status = ngx.status
+    if status > 499 and ctx.balancer_data then
+      local balancer_data = ctx.balancer_data
+      local try_count = balancer_data.try_count
+      local retries = balancer_data.retries
+      if try_count > retries then
+        local current_try = balancer_data.tries[try_count]
+        current_try.state = "failed"
+        current_try.code = status
+      end
+    end
+
+    local hash_cookie = ctx.balancer_data.hash_cookie
+    if hash_cookie then
+      balancer.set_cookie(hash_cookie)
+    end
+  end,
+  after = function(ctx)
+    local enabled_headers = kong.configuration.enabled_headers
+    local headers = constants.HEADERS
+    if ctx.KONG_PROXIED then
+      if enabled_headers[headers.UPSTREAM_LATENCY] then
+        header[headers.UPSTREAM_LATENCY] = ctx.KONG_WAITING_TIME
+      end
+
+      if enabled_headers[headers.PROXY_LATENCY] then
+        header[headers.PROXY_LATENCY] = ctx.KONG_PROXY_LATENCY
+      end
+
+      if enabled_headers[headers.VIA] then
+        header[headers.VIA] = SERVER_HEADER
+      end
+
+    else
+      if enabled_headers[headers.RESPONSE_LATENCY] then
+        header[headers.RESPONSE_LATENCY] = ctx.KONG_RESPONSE_LATENCY
+      end
+
+      -- Some plugins short-circuit the request with Via-header, and in those cases
+      -- we don't want to set the Server-header, if the Via-header matches with
+      -- the Kong server header.
+      if not (enabled_headers[headers.VIA] and header[headers.VIA] == SERVER_HEADER) then
+        if enabled_headers[headers.SERVER] then
+          header[headers.SERVER] = SERVER_HEADER
+
+        else
+          header[headers.SERVER] = nil
+        end
```

Comment:
Personal preference but I would write this to be more defensive
```suggestion
        header[headers.SERVER] = nil
        if enabled_headers[headers.SERVER] then
          header[headers.SERVER] = SERVER_HEADER
        end
```

Reply:
@colinh-kong I think it is not that more defensive and less optimized on this hot path. I have not measured it though. E.g. your code may set the header twice (and by default it does) while the previous code only set it once.